README.md
ipinformation
========

ipinformation is a python package focused on combining information about an IP in JSON format.

Features
========
* whois (via https://github.com/secynic/ipwhois)
* AS Name and Number (via https://www.team-cymru.com or local Maxmind database)
* reverse/ptr address
* geo location (via https://www.maxmind.com)
* version type
* conversion to bits (if you ever want to subnet search)
* type (public,private,netmask,etc).

Please Note
===========
* Converts all timestamps to UTC.
* Currently only supports IPv4.
* Requires internet access for: querying IP whois servers; performing dns lookup against cymru; dns lookups for reverse ptr; downloading maxmind databases

Requirements
============
* Python 2.7
* pip install -U requests[security] #Install requests security
* pip install -U pygeoip; #Install legacy maxmind geoip
* pip install -U netaddr; #Install network address tool
* pip install -U ipwhois; #Install ip whois tool
* pip install -U dateutils; #Time/Date Utility
* pip install -U dnspython; #DNS Utility

Install
=======
pip install -e git+https://github.com/neu5ron/ipinformation@master#egg=ipinformation

Usage Examples
==============
Valid IP
---------
	is_ip( ) = Return true if valid IP address return false if invalid IP address
	>>> from ipinformation import IPInformation
	>>> print IPInformation(ip_address='8.8.8.8').is_ip()
		True
	>>> print IPInformation(ip_address='NotAnIP').is_ip()
		False

Is Public IP
-------------
	is_public( ) = Return true if an IP address is publicly accessible/routable
	>>> from ipinformation import IPInformation
	>>> print IPInformation(ip_address='8.8.8.8').is_public()
		True
	>>> print IPInformation(ip_address='127.0.0.1').is_public()
		False

General Information
-------------------
	general_info( ) = Return IP in bits, ip_type (ie: private, multicast, loopback,etc..), time updated/returned and version for an IP Address
	>>> from ipinformation import IPInformation
	>>> from pprint import pprint
	>>> pprint( IPInformation(ip_address='8.8.8.8').general_info() )
	{'general': {'bits': '00001000000010000000100000001000',
				 'type': 'public',
				 'updated': datetime.datetime(2016, 1, 16, 18, 7, 4, 288512),
				 'version': '4'}}
	>>> IPInformation(ip_address='127.0.0.1').general_info()
	{'general': {'bits': '01111111000000000000000000000001',
				 'type': 'loopback',
				 'updated': datetime.datetime(2016, 1, 16, 18, 10, 6, 729149),
				 'version': '4'}}
Geo Information
---------------
	geo_info( ) = Return Geo location information (City,State,Country,etc...) for an IP Address
	>>> from ipinformation import IPInformation
	>>> from pprint import pprint
	>>> pprint( IPInformation(ip_address='8.8.8.8').geo_info() )
	{'geo': {'area_code': 650,
			 'city': u'Mountain View',
			 'continent': 'NA',
			 'coordinates': [37.3845, -122.0881],
			 'country_code': 'US',
			 'country_code3': 'USA',
			 'country_name': 'United States',
			 'dma_code': 807,
			 'error': 'no',
			 'latitude': 37.3845,
			 'longitude': -122.0881,
			 'metro_code': 'San Francisco, CA',
			 'postal_code': u'94040',
			 'region_code': u'CA',
			 'time_zone': 'America/Los_Angeles'}}
	>>> pprint( IPInformation(ip_address='127.0.0.1').geo_info() )
	{'geo': {'general': {'area_code': None,
						 'asname': None,
						 'asnum': None,
						 'city': None,
						 'continent': None,
						 'coordinates': None,
						 'country_code': None,
						 'country_code3': None,
						 'country_name': None,
						 'dma_code': None,
						 'error': 'no',
						 'latitude': None,
						 'longitude': None,
						 'metro_code': None,
						 'postal_code': None,
						 'region_code': None,
						 'time_zone': None}}}
Whois Information
-----------------
	whois_info( ) = Return WhoisInfo of the IP (AS Name/Number/CIDR/etc...,Subnet, CIDR, City,State,Country,Address, etc...) for an IP Address
	>>> from ipinformation import IPInformation
	>>> from pprint import pprint
	>>> pprint( IPInformation(ip_address='8.8.8.8').whois_info() )
	{'whois': {'as': {'cidr': '8.8.8.0/24',
					  'country_code': 'US',
					  'creation_date': None,
					  'name': u'Google Inc.',
					  'number': 15169,
					  'registry': 'arin'},
			   'error': 'no',
			   'raw': '\n#\n# ARIN WHOIS data and services are subject to the Terms of Use\n# available at: https://www.arin.net/whois_tou.html\n#\n# If you see inaccuracies in the results, please report at\n# http://www.arin.net/public/whoisinaccuracy/index.xhtml\n#\n\n\n#\n# The following results may also be obtained via:\n# http://whois.arin.net/rest/nets;q=8.8.4.4?showDetails=true&showARIN=false&showNonArinTopLevelNet=false&ext=netref2\n#\n\n\n# start\n\nNetRange:       8.0.0.0 - 8.255.255.255\nCIDR:           8.0.0.0/8\nNetName:        LVLT-ORG-8-8\nNetHandle:      NET-8-0-0-0-1\nParent:          ()\nNetType:        Direct Allocation\nOriginAS:       \nOrganization:   Level 3 Communications, Inc. (LVLT)\nRegDate:        1992-12-01\nUpdated:        2012-02-24\nRef:            http://whois.arin.net/rest/net/NET-8-0-0-0-1\n\n\n\nOrgName:        Level 3 Communications, Inc.\nOrgId:          LVLT\nAddress:        1025 Eldorado Blvd.\nCity:           Broomfield\nStateProv:      CO\nPostalCode:     80021\nCountry:        US\nRegDate:        1998-05-22\nUpdated:        2012-01-30\nComment:        ADDRESSES WITHIN THIS BLOCK ARE NON-PORTABLE\nRef:            http://whois.arin.net/rest/org/LVLT\n\n\nOrgTechHandle: IPADD5-ARIN\nOrgTechName:   ipaddressing\nOrgTechPhone:  +1-877-453-8353 \nOrgTechEmail:  ipaddressing@level3.com\nOrgTechRef:    http://whois.arin.net/rest/poc/IPADD5-ARIN\n\nOrgNOCHandle: NOCSU27-ARIN\nOrgNOCName:   NOC Support\nOrgNOCPhone:  +1-877-453-8353 \nOrgNOCEmail:  noc.coreip@level3.com\nOrgNOCRef:    http://whois.arin.net/rest/poc/NOCSU27-ARIN\n\nOrgAbuseHandle: APL8-ARIN\nOrgAbuseName:   Abuse POC LVLT\nOrgAbusePhone:  +1-877-453-8353 \nOrgAbuseEmail:  abuse@level3.com\nOrgAbuseRef:    http://whois.arin.net/rest/poc/APL8-ARIN\n\n# end\n\n\n# start\n\nNetRange:       8.8.4.0 - 8.8.4.255\nCIDR:           8.8.4.0/24\nNetName:        LVLT-GOGL-8-8-4\nNetHandle:      NET-8-8-4-0-1\nParent:         LVLT-ORG-8-8 (NET-8-0-0-0-1)\nNetType:        Reallocated\nOriginAS:       \nOrganization:   Google Inc. (GOGL)\nRegDate:        2014-03-14\nUpdated:        2014-03-14\nRef:            http://whois.arin.net/rest/net/NET-8-8-4-0-1\n\n\n\nOrgName:        Google Inc.\nOrgId:          GOGL\nAddress:        1600 Amphitheatre Parkway\nCity:           Mountain View\nStateProv:      CA\nPostalCode:     94043\nCountry:        US\nRegDate:        2000-03-30\nUpdated:        2015-11-06\nRef:            http://whois.arin.net/rest/org/GOGL\n\n\nOrgAbuseHandle: ABUSE5250-ARIN\nOrgAbuseName:   Abuse\nOrgAbusePhone:  +1-650-253-0000 \nOrgAbuseEmail:  network-abuse@google.com\nOrgAbuseRef:    http://whois.arin.net/rest/poc/ABUSE5250-ARIN\n\nOrgTechHandle: ZG39-ARIN\nOrgTechName:   Google Inc\nOrgTechPhone:  +1-650-253-0000 \nOrgTechEmail:  arin-contact@google.com\nOrgTechRef:    http://whois.arin.net/rest/poc/ZG39-ARIN\n\n# end\n\n\n\n#\n# ARIN WHOIS data and services are subject to the Terms of Use\n# available at: https://www.arin.net/whois_tou.html\n#\n# If you see inaccuracies in the results, please report at\n# http://www.arin.net/public/whoisinaccuracy/index.xhtml\n#\n\n',
			   'registration': [{'abuse_emails': None,
								 'address': '1025 Eldorado Blvd.',
								 'cidr': '8.0.0.0/8',
								 'city': 'Broomfield',
								 'country_code': 'US',
								 'creation_date': datetime.datetime(1992, 12, 1, 0, 0, tzinfo=<UTC>),
								 'description': 'Level 3 Communications, Inc.',
								 'handle': 'NET-8-0-0-0-1',
								 'misc_emails': None,
								 'name': 'LVLT-ORG-8-8',
								 'postal_code': '80021',
								 'range': '8.0.0.0-8.255.255.255',
								 'state': 'CO',
								 'tech_emails': None,
								 'updated': datetime.datetime(2012, 2, 24, 0, 0, tzinfo=<UTC>)},
								{'abuse_emails': None,
								 'address': '1600 Amphitheatre Parkway',
								 'cidr': '8.8.8.0/24',
								 'city': 'Mountain View',
								 'country_code': 'US',
								 'creation_date': datetime.datetime(2014, 3, 14, 0, 0, tzinfo=<UTC>),
								 'description': 'Google Inc.',
								 'handle': 'NET-8-8-8-0-1',
								 'misc_emails': None,
								 'name': 'LVLT-GOGL-8-8-8',
								 'postal_code': '94043',
								 'range': '8.8.8.0-8.8.8.255',
								 'state': 'CA',
								 'tech_emails': None,
								 'updated': datetime.datetime(2014, 3, 14, 0, 0, tzinfo=<UTC>)}],
			   }}
	>>> pprint( IPInformation(ip_address='127.0.0.1').whois_info() )
	No Whois information for '127.0.0.1' because it is not a public ip

	{'whois': {'as': {'cidr': None,
					  'country_code': None,
					  'creation_date': None,
					  'name': None,
					  'number': None,
					  'registry': None},
			   'raw': None,
			   'registration': [{'abuse_emails': None,
								 'address': None,
								 'cidr': None,
								 'city': None,
								 'country_code': None,
								 'creation_date': None,
								 'description': None,
								 'handle': None,
								 'misc_emails': None,
								 'postal_code': None,
								 'state': None,
								 'tech_emails': None,
								 'updated': None}],
			   }}
	"""

Gather AS Name and Number from local Maxmind Database
-----------------------------------------------------
        maxmind_AS ( ) = Use Maxmind DB to gather ASName and ASNumber instead of doing a network lookup for it.
        >>> from ipinformation import IPInformation
        >>> from pprint import pprint
        >>> pprint( IPInformation(ip_address='8.8.8.8').maxmind_AS() )
        {'as': {'name': 'Google Inc.', 'number': 15169}}
Gather AS Information from Cymru
-----------------------------------------
        cymru_AS ( ) = Use Team Cymru "$IPReversed.origin.asn.cymru.com' to gather ASName and ASNumber via DNS Lookup
        >>> from ipinformation import IPInformation
        >>> from pprint import pprint
        >>> pprint( IPInformation(ip_address='8.8.8.8').cymru_AS() )
        {'as': {'cidr': '8.8.8.0/24',
                'country_code': 'US',
                'creation_date': '',
                'number': 15169,
                'registry': 'arin'}}
                'error': False}}

Gather PTR DNS Record
-----------------------------------------
        get_ptr ( ) = Get the PTR address of an IP address. Reverse IP.
        :param DNSServer: str -- of optional DNS server to use. Otherwise uses system DNS
        :param DNSPort: int -- of port to use on the server. Requires DNSServer. Default = 53
        :param DNSTimeout: int/float -- of timeout in seconds. Default = 4
        :param DNSTCP: bool -- choose True to use TCP otherwise defaults to UDP.
        >>> from ipinformation import IPInformation
        >>> from pprint import pprint
        >>> pprint( IPInformation(ip_address='8.8.8.8').get_ptr() )
        {'error': None, 'ptr_address': 'google-public-dns-a.google.com'}
        :return: dict

All Information / Put everything together
-----------------------------------------
	all( ) = Return general, geo, and whois information for an IP Address
	>>> from ipinformation import IPInformation
	>>> from pprint import pprint
	>>> pprint( IPInformation(ip_address='8.8.8.8').all() )
	{'general': {'bits': '00001000000010000000100000001000',
				 'type': 'public',
				 'updated': datetime.datetime(2016, 1, 16, 18, 26, 23, 487181),
				 'version': '4'},
	 'geo': {'area_code': 650,
			 'city': u'Mountain View',
			 'continent': 'NA',
			 'coordinates': [37.3845, -122.0881],
			 'country_code': 'US',
			 'country_code3': 'USA',
			 'country_name': 'United States',
			 'dma_code': 807,
			 'error': 'no',
			 'latitude': 37.3845,
			 'longitude': -122.0881,
			 'metro_code': 'San Francisco, CA',
			 'postal_code': u'94040',
			 'region_code': u'CA',
			 'time_zone': 'America/Los_Angeles'},
	 'whois': {'as': {'cidr': '8.8.8.0/24',
					  'country_code': 'US',
					  'creation_date': None,
					  'name': u'Google Inc.',
					  'number': 15169,
					  'registry': 'arin'},
			   'error': 'no',
			   'raw': '\n#\n# ARIN WHOIS data and services are subject to the Terms of Use\n# available at: https://www.arin.net/whois_tou.html\n#\n# If you see inaccuracies in the results, please report at\n# http://www.arin.net/public/whoisinaccuracy/index.xhtml\n#\n\n\n#\n# The following results may also be obtained via:\n# http://whois.arin.net/rest/nets;q=8.8.4.4?showDetails=true&showARIN=false&showNonArinTopLevelNet=false&ext=netref2\n#\n\n\n# start\n\nNetRange:       8.0.0.0 - 8.255.255.255\nCIDR:           8.0.0.0/8\nNetName:        LVLT-ORG-8-8\nNetHandle:      NET-8-0-0-0-1\nParent:          ()\nNetType:        Direct Allocation\nOriginAS:       \nOrganization:   Level 3 Communications, Inc. (LVLT)\nRegDate:        1992-12-01\nUpdated:        2012-02-24\nRef:            http://whois.arin.net/rest/net/NET-8-0-0-0-1\n\n\n\nOrgName:        Level 3 Communications, Inc.\nOrgId:          LVLT\nAddress:        1025 Eldorado Blvd.\nCity:           Broomfield\nStateProv:      CO\nPostalCode:     80021\nCountry:        US\nRegDate:        1998-05-22\nUpdated:        2012-01-30\nComment:        ADDRESSES WITHIN THIS BLOCK ARE NON-PORTABLE\nRef:            http://whois.arin.net/rest/org/LVLT\n\n\nOrgTechHandle: IPADD5-ARIN\nOrgTechName:   ipaddressing\nOrgTechPhone:  +1-877-453-8353 \nOrgTechEmail:  ipaddressing@level3.com\nOrgTechRef:    http://whois.arin.net/rest/poc/IPADD5-ARIN\n\nOrgNOCHandle: NOCSU27-ARIN\nOrgNOCName:   NOC Support\nOrgNOCPhone:  +1-877-453-8353 \nOrgNOCEmail:  noc.coreip@level3.com\nOrgNOCRef:    http://whois.arin.net/rest/poc/NOCSU27-ARIN\n\nOrgAbuseHandle: APL8-ARIN\nOrgAbuseName:   Abuse POC LVLT\nOrgAbusePhone:  +1-877-453-8353 \nOrgAbuseEmail:  abuse@level3.com\nOrgAbuseRef:    http://whois.arin.net/rest/poc/APL8-ARIN\n\n# end\n\n\n# start\n\nNetRange:       8.8.4.0 - 8.8.4.255\nCIDR:           8.8.4.0/24\nNetName:        LVLT-GOGL-8-8-4\nNetHandle:      NET-8-8-4-0-1\nParent:         LVLT-ORG-8-8 (NET-8-0-0-0-1)\nNetType:        Reallocated\nOriginAS:       \nOrganization:   Google Inc. (GOGL)\nRegDate:        2014-03-14\nUpdated:        2014-03-14\nRef:            http://whois.arin.net/rest/net/NET-8-8-4-0-1\n\n\n\nOrgName:        Google Inc.\nOrgId:          GOGL\nAddress:        1600 Amphitheatre Parkway\nCity:           Mountain View\nStateProv:      CA\nPostalCode:     94043\nCountry:        US\nRegDate:        2000-03-30\nUpdated:        2015-11-06\nRef:            http://whois.arin.net/rest/org/GOGL\n\n\nOrgAbuseHandle: ABUSE5250-ARIN\nOrgAbuseName:   Abuse\nOrgAbusePhone:  +1-650-253-0000 \nOrgAbuseEmail:  network-abuse@google.com\nOrgAbuseRef:    http://whois.arin.net/rest/poc/ABUSE5250-ARIN\n\nOrgTechHandle: ZG39-ARIN\nOrgTechName:   Google Inc\nOrgTechPhone:  +1-650-253-0000 \nOrgTechEmail:  arin-contact@google.com\nOrgTechRef:    http://whois.arin.net/rest/poc/ZG39-ARIN\n\n# end\n\n\n\n#\n# ARIN WHOIS data and services are subject to the Terms of Use\n# available at: https://www.arin.net/whois_tou.html\n#\n# If you see inaccuracies in the results, please report at\n# http://www.arin.net/public/whoisinaccuracy/index.xhtml\n#\n\n',
			   'registration': [{'abuse_emails': None,
								 'address': '1025 Eldorado Blvd.',
								 'cidr': '8.0.0.0/8',
								 'city': 'Broomfield',
								 'country_code': 'US',
								 'creation_date': datetime.datetime(1992, 12, 1, 0, 0, tzinfo=<UTC>),
								 'description': 'Level 3 Communications, Inc.',
								 'handle': 'NET-8-0-0-0-1',
								 'misc_emails': None,
								 'name': 'LVLT-ORG-8-8',
								 'postal_code': '80021',
								 'range': '8.0.0.0-8.255.255.255',
								 'state': 'CO',
								 'tech_emails': None,
								 'updated': datetime.datetime(2012, 2, 24, 0, 0, tzinfo=<UTC>)},
								{'abuse_emails': None,
								 'address': '1600 Amphitheatre Parkway',
								 'cidr': '8.8.8.0/24',
								 'city': 'Mountain View',
								 'country_code': 'US',
								 'creation_date': datetime.datetime(2014, 3, 14, 0, 0, tzinfo=<UTC>),
								 'description': 'Google Inc.',
								 'handle': 'NET-8-8-8-0-1',
								 'misc_emails': None,
								 'name': 'LVLT-GOGL-8-8-8',
								 'postal_code': '94043',
								 'range': '8.8.8.0-8.8.8.255',
								 'state': 'CA',
								 'tech_emails': None,
								 'updated': datetime.datetime(2014, 3, 14, 0, 0, tzinfo=<UTC>)}],
			   }}
	>>> pprint( IPInformation(ip_address='127.0.0.1').all() )
	No Whois information for '127.0.0.1' because it is not a public ip

	{'general': {'bits': '01111111000000000000000000000001',
				 'type': 'loopback',
				 'updated': datetime.datetime(2016, 1, 16, 18, 27, 41, 528938),
				 'version': '4'},
	 'geo': {'general': {'area_code': None,
						 'asname': None,
						 'asnum': None,
						 'city': None,
						 'continent': None,
						 'coordinates': None,
						 'country_code': None,
						 'country_code3': None,
						 'country_name': None,
						 'dma_code': None,
						 'error': 'no',
						 'latitude': None,
						 'longitude': None,
						 'metro_code': None,
						 'postal_code': None,
						 'region_code': None,
						 'time_zone': None}},
	 'whois': {'as': {'cidr': None,
					  'country_code': None,
					  'creation_date': None,
					  'name': None,
					  'number': None,
					  'registry': None},
			   'raw': None,
			   'registration': [{'abuse_emails': None,
								 'address': None,
								 'cidr': None,
								 'city': None,
								 'country_code': None,
								 'creation_date': None,
								 'description': None,
								 'handle': None,
								 'misc_emails': None,
								 'postal_code': None,
								 'state': None,
								 'tech_emails': None,
								 'updated': None}],
			   }}

CHANGES.md
Changelog
=========
1.0.21 (2016-08-11)
-------------------
-forgot to strip trailing dot in ptr

1.0.20 (2016-08-01)
-------------------
-optimization...

1.0.19 (2016-08-01)
-------------------
-fixes

1.0.18 (2016-07-20)
-------------------
-added better checking for IPv4
-added ability to get PTR record for IP

1.0.17 (2016-07-06)
-------------------
-added Team Cymru AS documentation

1.0.16 (2016-06-11)
-------------------
-fix as

1.0.15 (2016-06-11)
-------------------
-none error :(

1.0.14 (2016-06-11)
-------------------
-Added JSON 'error' if AS was successful or not

1.0.13 (2016-06-11)
-------------------
-Except AttributeError if unable to find AS in Maxmind DB
-Added Team Cymru AS Lookup

1.0.12 (2016-06-03)
-------------------
-Except AttributeError if unable to find AS in Maxmind DB

1.0.11 (2016-04-10)
-------------------
-Log fix

1.0.10 (2016-03-29)
-------------------
-Changed to lookup_whois function for ipwhois as old one is going to eventually be removed

1.0.9 (2016-03-22)
-------------------
-Some IPs returning none for maxmind geoip

1.0.8 (2016-03-17)
-------------------
-FML...array geo-points are [longitude,latitude]—the opposite order!

1.0.7 (2016-03-14)
-------------------
-With ipwhois version 0.11.2 allows to not use whois for cyrmu and only dns in order to not get blacklisted and if it fails.
-Set Maxmind DBs to only update second tuesday of every month. Technically just looks to see if its greater than 14 days since download
-Added ability to grab AS Name & Number from Maxmind DB instead of having to do an online lookup
-Added logging and removed prints
-Using HTTPs maxmind database now

1.0.6 (2016-01-27)
-------------------
-Using with open file instead of just open.

1.0.5 (2016-01-26)
-------------------
-Added whois raw information

1.0.4 (2016-01-18)
-------------------
-Minor changes

1.0.3 (2016-01-18)
-------------------
-Fixed home directory for windows or linux

1.0.2 (2016-01-18)
-------------------
-Added pprint for examples

1.0.1 (2016-01-18)
-------------------
-GD unicode returns from MongoDB
-removed some unnecessary print newlines

1.0.0 (2016-01-18)
-------------------
-initial commit

ipinformation/GeoDBConnection.py
#!/usr/bin/env python2.7
#-*- coding: utf-8 -*-
from datetime import datetime
import os, sys
import requests
import gzip
import io
import pygeoip
from dateutil import parser
import logging
from logging import handlers

######################################## # Edit If Need Be
base_directory = os.path.expanduser("~") #Base directory where 'GeoIP' directory will be created (currently home directory)
hours_to_pull_new_geoip_db = 336 #Use this variable in hours to determine how often to download and update the local databases. Default is 14 days
geoip_directory = os.path.join( base_directory, 'GeoIP' )
########################################

# Set logging
log_file = os.path.join ( os.path.join( '/tmp', 'ipinformation.log' ) )
logging_file = logging.getLogger(__name__)
logging_file.setLevel(logging.DEBUG)
logging_file_handler = handlers.RotatingFileHandler( log_file, maxBytes=5, backupCount=0  )
info_format = logging.Formatter('%(asctime)s - %(filename)s - %(levelname)s - Function: %(funcName)s - LineNumber: %(lineno)s - %(message)s')
logging_file_handler.setFormatter(info_format)
logging_file.addHandler(logging_file_handler)


if not os.path.exists(geoip_directory):
    # print '%s does not exist. Creating it now.\n'%geoip_directory
    logging_file.info( '{0} does not exist. Creating it now.'.format(geoip_directory) )

    try:
        os.mkdir(geoip_directory)

    except OSError as error:
        print 'Failed to create {0}. Due to:\n{1}'.format( geoip_directory, error )
        logging_file.error( 'Failed to create {0}. Due to:\n{1}'.format( geoip_directory, error ) )
        sys.exit(1)


class GeoIPDB():
    #Doucmentation: https://pythonhosted.org/pygeoip/
    def __init__(self):
        pass

    def geoipv4_country(self):
        return self.download_and_update_geoip( 'GeoIPv4_Country.dat', 'geoipv4_country', 'https://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz' )

    def geoipv6_country(self):
        return self.download_and_update_geoip( 'GeoIPv6_Country.dat', 'geoipv6_country', 'https://geolite.maxmind.com/download/geoip/database/GeoIPv6.dat.gz' )

    def geoipv4_city(self):
        return self.download_and_update_geoip( 'GeoIPv4_City.dat', 'geoipv4_city', 'https://geolite.maxmind.com/download/geoip/database/GeoLiteCity.dat.gz' )

    def geoipv6_city(self):
       return self.download_and_update_geoip( 'GeoIPv6_City.dat', 'geoipv6_city', 'https://geolite.maxmind.com/download/geoip/database/GeoLiteCityv6-beta/GeoLiteCityv6.dat.gz' )

    def geoipv4_as(self):
        return self.download_and_update_geoip( 'GeoIP_ASv4.dat', 'geoipv4_as', 'https://download.maxmind.com/download/geoip/database/asnum/GeoIPASNum.dat.gz' )

    def geoipv6_as(self):
        return self.download_and_update_geoip( 'GeoIP_ASv6.dat', 'geoipv6_as', 'https://download.maxmind.com/download/geoip/database/asnum/GeoIPASNumv6.dat.gz' )

    def download_and_update_geoip( self, filename, variable_name, download_url ):
        """
        Update Or Download GeoIP information from MaxMind and return it for use
        :param filename:
        :param variable_name:
        :param download_url:
        :return:
        """

        current_time = datetime.utcnow()
        need_to_download = False
        file_last_downloaded = os.path.join( geoip_directory, 'last_downloaded_%s.txt'%filename ) # File that will be used to determine the last time the DBs were downloaded

        if os.path.exists( os.path.join( geoip_directory, filename ) ):

            # Check to see if download timestamp exists and if it does see time diff since download
            if os.path.exists(file_last_downloaded):
                with open( file_last_downloaded, 'r' ) as f:
                    last_downloaded = f.read().strip()
                f.close()
                time_diff = (current_time - parser.parse(last_downloaded)).total_seconds()
                if time_diff > hours_to_pull_new_geoip_db*3600:
                    need_to_download = True

            else:
                # Set download timestamp if it was not downloaded using this script and this is the first time the script is ran
                with open( file_last_downloaded, 'w+' ) as f:
                    f.write(str(current_time))
                f.close()

        else:
            need_to_download = True

        if need_to_download:
            print '%s file needs to be updated or does not exist!\nTrying to download it to %s/%s\n'%( variable_name, geoip_directory, filename )
            logging_file.info( '{0} file needs to be updated or does not exist! Trying to download it to "{1}/{2}"'.format( variable_name, geoip_directory, filename ) )
            try:
                requests.packages.urllib3.disable_warnings()
                response = requests.get( download_url,timeout=(10,2), verify=False )
                compressed_content = io.BytesIO(response.content)
                decompressed_content = gzip.GzipFile(fileobj=compressed_content)
                with open( os.path.join( geoip_directory, filename ), 'wb' ) as downloaded_file:
                    downloaded_file.write(decompressed_content.read())
                    with open( file_last_downloaded, 'w+' ) as f:
                        f.write(str(current_time))
                    f.close()
                downloaded_file.close()

            except IOError as error:
                print 'Could not download and write GeoIP database. Due to:\n{0}'.format( error )
                logging_file.error( 'Could not download and write GeoIP database. Due to:\n{0}'.format( error ) )
                sys.exit(1)

            except requests.HTTPError as error:
                print 'Could not download and write GeoIP database. Due to:\n{0}'.format( error )
                logging_file.error( 'Could not download and write GeoIP database. Due to:\n{0}'.format( error ) )
                sys.exit(1)

            except requests.Timeout as error:
                print 'Could not download and write GeoIP database. Due to:\n{0}'.format( error )
                logging_file.error( 'Could not download and write GeoIP database. Due to:\n{0}'.format( error ) )
                sys.exit(1)

            except requests.TooManyRedirects as error:
                print 'Could not download and write GeoIP database. Due to:\n{0}'.format( error )
                logging_file.error( 'Could not download and write GeoIP database. Due to:\n{0}'.format( error ) )
                sys.exit(1)

            except requests.ConnectionError as error:
                print 'Could not download and write GeoIP database. Due to:\n{0}'.format( error )
                logging_file.error( 'Could not download and write GeoIP database. Due to:\n{0}'.format( error ) )
                sys.exit(1)

        return pygeoip.GeoIP( os.path.join( geoip_directory, filename ) )

ipinformation/time_info.py
#!/usr/bin/env python2.7
#-*- coding: utf-8 -*-
from dateutil import parser
from pytz import timezone, UTC, exceptions, FixedOffset
from datetime import datetime
import sys

# http://www.worldtimeserver.com/current_time_in_US-NY.aspx#
time_ordinals = [ 32, 43, 45, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58 ]

timezone_deltas = [ '+0000', '+0100', '+0200', '+0300', '+0400', '+0500', '+0600', '+0700', '+0800', '+0900', '+1000', '+1100', '+1200', '-0000', '-0100', '-0200', '-0300', '-0400', '-0500', '-0600', '-0700', '-0800', '-0900', '-1000', '-1100', '-1200' ]
#Should ever use offsets '+1300' or '+1400' ?

class convert_time(object):
    def __init__( self, time_to_convert ):
        self.time_to_convert = time_to_convert
        if not time_to_convert:
            self.time_to_convert = None

    def convert_to_utc( self, utc_offset='+0000', is_dst=True  ):
        """convert_to_utc( ) = Converts a timestamp to UTC/GMT.
        If the timestamp has the timezone offset than it will use that instead of the utc_offset variable passed.
        >>> from ipinformation import time_info
        >>> non_offset_aware_time = '2014/09/26 17:05:22'
        >>> tz = '-0500'
        >>> utc_time = time_info.convert_time(non_offset_aware_time).convert_to_utc(utc_offset=tz)
        Original Time: 2014/09/26 17:05:22
        UTC Time: 2014-09-26 22:05:22+00:00
        >>> from ipinformation import time_info
        >>> offset_aware_time = '2014/09/26 17:05:22-0500'
        >>> utc_time = time_info.convert_time(non_offset_aware_time).convert_to_utc()
        Original Time: 2014/09/26 17:05:22
        UTC Time: 2014-09-26 22:05:22+00:00
        """
        #TODO:Make OffSet Dynamic and With DayLightSavings or Not
        if self.time_to_convert is None:
            return None

        date_str =  parser.parse(self.time_to_convert)

        # Timezone not found convert to UTC with offset passed
        if date_str.tzinfo is None:
            # Return if already UTC
            if utc_offset == '+0000' or utc_offset == '-0000':
                return timezone('UTC').localize( date_str, is_dst )
            tz = self.convert_offset_to_timezone_aware(utc_offset)
            # Invert timezone to convert to appropriate UTC time
            tz_num = tz.replace('Etc/GMT','')
            if tz_num[0] == '-': #Convert to positive offset
                tz_flipped = 'Etc/GMT+%s'%tz_num[1:]
            else: #Convert to negative offset:
                tz_flipped = 'Etc/GMT-%s'%tz_num[1:]
            try:
                date_str = timezone(tz_flipped).localize(date_str, is_dst)
                return timezone('UTC').normalize( date_str, is_dst )
            except (ValueError, exceptions.UnknownTimeZoneError) as e:
                print '%s for %s with variable utc_offset:%s\n'%(e,self.time_to_convert,utc_offset)
                # print 'Try one of these timezone offsets/deltas%s\n'%timezone_deltas
                sys.exit(0)
        # Timezone found in timestamp use that instead
        else:
            return timezone('UTC').normalize( date_str, is_dst )


    def time_in_between( self, first_time, is_utc=False ):
        """time_in_between( ) = returns the days in between two timestamps"""
        # time_to_convert = def utc_conversion( self, utc_offset=None  )#TODO:Maybe use this for non utc time_in_between
        days_in_between = (self.time_to_convert - first_time).days
        # if is_utc == 'yes':#TODO:use or delete...Need to use in case ever used with non offset aware or non parsed timestamp
        #     print self.time_to_convert, first_time
        #     days_in_between = ( self.time_to_convert - first_time ).days
        #
        # else:
        #     utc = pytz.UTC
        #     days_in_between = ( self.time_to_convert - utc.localize(first_time) ).days
        return days_in_between

    def convert_offset_to_timezone_aware(self, tz_number):
        """convert_offset_to_timezone_aware( ) = take an offset and make it into the pyzt aware GMT timezone"""
        if tz_number in timezone_deltas:
            if tz_number[1:3] == '00':
                tz_info = 'Etc/GMT0'
                return tz_info
            elif tz_number[1] == '0':
                tz_info = 'Etc/GMT%s'%tz_number[0:3].replace('0','')
                return tz_info
            else:
                tz_info = 'Etc/GMT%s'%tz_number[0:3]
                return tz_info
        else:
            print '%s is an Unknown UTC offset/delta\n'%tz_number
            print 'Try one of these timezone offsets/deltas%s\n'%timezone_deltas
            sys.exit(0)